### PR TITLE
[io] Teach TMemFile to work with regions with no extra memory copy

### PR DESCRIPTION
**Pull request by vgvassilev**
This is useful for moving the rdict files into the cxxmodule file as an extension. The relevant PR is coming soon.

### CODE REVIEW

File: io/io/src/TMemFile.cxx [hunk 1]
```diff
@@ -134,13 +134,19 @@ TMemFile::TMemFile(const char *path, ExternalDataPtr_t data) :
    Init(/* create */ false);
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Constructor to create a TMemFile re-using external storage.
+
+TMemFile::TMemFile(const char *path, ExternalDataPtr_t data)
```
Comment (pcanal) [inline io/io/src/TMemFile.cxx]: This seems to significantly change the ownership semantic of this signature (taking an std::shared_ptr).  What am I missing?

Comment (vgvassilev) [reply io/io/src/TMemFile.cxx]: How so? This reuses code by delegating to the other non-owning constructor. The ownership remains the same and is managed by the shared_ptr if the user chose that interface.

Comment (pcanal) [reply io/io/src/TMemFile.cxx]: I see that 
```
   const ExternalDataPtr_t fExternalData; ///< shared file data / content
```
and replaced by
```
   Bool_t       fExternalData;            ///< If there is external data / content
```
I do not see where the new version can increment and later decrement the use count in the user's shared ptr.

Comment (vgvassilev) [reply io/io/src/TMemFile.cxx]: Ah, I see. Let me rethink that part.

File: io/io/src/TMemFile.cxx [hunk 2]
```diff
@@ -113,12 +113,12 @@ TMemFile::EMode TMemFile::ParseOption(Option_t *option)
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// Constructor to create a TMemFile re-using external storage.
+/// Constructor to create a TMemFile re-using external C-Style storage.
 
-TMemFile::TMemFile(const char *path, ExternalDataPtr_t data) :
-   TFile(path, "WEB", "read-only TMemFile", 0 /*compress*/),
-   fBlockList(reinterpret_cast<UChar_t*>(const_cast<char*>(data->data())), data->size()),
-   fExternalData(std::move(data)), fSize(fExternalData->size()), fSysOffset(0), fBlockSeek(nullptr), fBlockOffset(0)
```
Comment (vgvassilev) [inline io/io/src/TMemFile.cxx]: @Axel-Naumann, why we do `std::move(data)`? Isn't that making nullptr the calling pointer?

Comment (Axel-Naumann) [reply io/io/src/TMemFile.cxx]: But `data` is a copy of whatever the caller passed: no `const &`.

**Comment by vgvassilev** [comment]
@phsft-bot build just on windows10/default

**Comment by vgvassilev** [comment]
@phsft-bot build on ROOT-performance-centos7-multicore/default, ROOT-fedora27/noimt, ROOT-fedora29/python3, ROOT-ubuntu16/rtcxxmod, mac1014/cxx17, windows10/default with flags -DCTEST_TEST_EXCLUDE_NONE=On

File: io/io/inc/TMemFile.h [hunk 3]
```diff
@@ -39,7 +45,8 @@ class TMemFile : public TFile {
       Long64_t   fSize;
    };
    TMemBlock    fBlockList;               ///< Collection of memory blocks of size fgDefaultBlockSize
-   const ExternalDataPtr_t fExternalData; ///< shared file data / content
+   ExternalDataPtr_t fExternalData; ///< shared file data / content
```
Comment (pcanal) [inline io/io/inc/TMemFile.h]:
```suggestion
   ExternalDataPtr_t fExternalData;       ///< shared file data / content
```

File: io/io/inc/TMemFile.h [hunk 4]
```diff
@@ -19,8 +19,14 @@
...
    using ExternalDataPtr_t = std::shared_ptr<const std::vector<char>>;
+   /// A read-only memory range which we do not control.
+   struct ExternalDataRange_t {
+      const char *fStart;
+      const size_t fSize;
+      ExternalDataRange_t(const char * start, const size_t size) : fStart(start), fSize(size) {}
+   };
 
-private:
+protected:
```
Comment (pcanal) [inline io/io/inc/TMemFile.h]: Why does it need to become protected?

Comment (vgvassilev) [reply io/io/inc/TMemFile.h]: See the test case, it relies on checking if the internal  pointer is same as the one in the blob. To do that we need to access that data and we cannot do it if it is private.

**Comment by pcanal** [comment]
Can you look at the result of clang-tidy and clang-tidy-modernize?  On the surface they 'might' be relevant errors.

File: io/io/inc/TMemFile.h (hunk 4 above)
Comment (Axel-Naumann) [reply io/io/inc/TMemFile.h]: You could use `TDataMember`.

**Comment by vgvassilev** [comment]
@pcanal, the errors from clang-tidy-modernize are in the source code of gtest. However, the error from clang-tidy is valid: it is from existing code which I moved and did not touch. I do not know how to fix it, I suspect @Axel-Naumann would be a better person to address this...

**Comment by vgvassilev** [comment]
A proof that the problem exists before this PR is:
```
 cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=ON   -DCMAKE_BUILD_TYPE="Debug" -DLLVM_BUILD_TYPE="Debug" -Dbuiltin_clang="On" -Dbuiltin_gsl="On" -Dbuiltin_llvm="On" -Dpython="On" -Droottest="On" -Druntime_cxxmodules="On" -Dx11="Off" /home/performance-test/src_master 
-- ROOT default compression algorithm: zlib
-- Found GCC. Major version 7, minor version 4
-- Found a 64bit system
-- Found GNU compiler collection
-- ROOT Platform: linux
-- ROOT Architecture: linuxx8664gcc
-- Build Type: Debug
-- Compiler Flags:  -std=c++11 -Wno-implicit-fallthrough -Wno-noexcept-type -pipe  -Wshadow -Wall -W -Woverloaded-virtual -fsigned-char -pthread 
-- Looking for ZLib
-- Looking for Freetype
-- Looking for LZMA
-- Found LZMA includes at /usr/include
-- Found LZMA library at /usr/lib/x86_64-linux-gnu/liblzma.so
-- Building AfterImage library included in ROOT itself
-- Looking for GSL
-- Downloading and building GSL version 2.1
-- Looking for python
-- Enabled support for:  builtin_afterimage builtin_clang builtin_ftgl builtin_glew builtin_gsl builtin_llvm builtin_lz4 builtin_openssl builtin_pcre builtin_tbb builtin_vdt builtin_xxhash clad exceptions explicitlink gdml http imt mathmore mlp python roofit runtime_cxxmodules shared ssl tmva spectrum vdt
-- Native target architecture is X86
-- Threads disabled.
-- Doxygen disabled.
-- Go bindings disabled.
-- LLVM host triple: x86_64-unknown-linux-gnu
-- LLVM default target triple: x86_64-unknown-linux-gnu
-- Building with -fPIC
-- Constructing LLVMBuild project information
-- Linker detection: GNU ld
-- SysLibs: rt;dl;tinfo;-lpthread;/usr/lib/x86_64-linux-gnu/libz.so
-- Targeting X86
-- Clang version: 5.0.0
-- Cling version (from VERSION file): ROOT_0.6~dev
-- Cling will look for C++ headers in '/usr/include/c++/7:/usr/include/x86_64-linux-gnu/c++/7:/usr/include/c++/7/backward' at runtime.
-- And then fallback to: 'x86_64-linux-gnu-g++-7'
Recording the git revision now
-- Performing Test found_stdstringview
-- Performing Test found_stdstringview - Failed
-- Performing Test found_stdexpstringview
-- Performing Test found_stdexpstringview - Failed
-- Performing Test found_stdapply
-- Performing Test found_stdapply - Failed
-- Performing Test found_stdinvoke
-- Performing Test found_stdinvoke - Failed
-- Performing Test found_stdindexsequence
-- Performing Test found_stdindexsequence - Failed
-- 315/786 tutorials have been vetoed for various reasons
-- Using directly ROOTSYS for roottest.git is under deprecation, please add ROOT in PATH.
-- Check for bitness: Found 64 bit architecture.
-- Scanning subdirectories for tests...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/performance-test/build_master
performance-test@vv-nuc ~/build_master $ clang-tidy-7 '-checks=-*,clang-analyzer-*,-clang-analyzer-alpha*,bugprone*' ../src_master/io/io/src/TMemFile.cxx 
103 warnings generated.
/home/performance-test/src_master/io/io/src/TMemFile.cxx:152:4: error: the value returned by this function should be used [bugprone-unused-return-value,-warnings-as-errors]
   buffer.release();
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:152:4: note: cast the expression to void to silence this warning
/home/performance-test/src_master/io/io/src/TMemFile.cxx:188:40: error: Value stored to 'mode' is never read [clang-analyzer-deadcode.DeadStores,-warnings-as-errors]
      if (optmode == EMode::kRecreate) mode |= O_TRUNC;
                                       ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:188:40: note: Value stored to 'mode' is never read
/home/performance-test/src_master/io/io/src/TMemFile.cxx:190:12: error: Call to virtual function during construction [clang-analyzer-optin.cplusplus.VirtualCall,-warnings-as-errors]
      fD = SysOpen(path, O_RDWR | O_CREAT, 0644);
           ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:186:4: note: Taking true branch
   if (NeedsToWrite(optmode)) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:188:7: note: Taking false branch
      if (optmode == EMode::kRecreate) mode |= O_TRUNC;
      ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:190:12: note: This constructor of an object of type 'TMemFile' has not returned when the virtual method was called
      fD = SysOpen(path, O_RDWR | O_CREAT, 0644);
           ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:190:12: note: Call to virtual function during construction
/home/performance-test/src_master/io/io/src/TMemFile.cxx:198:12: error: Call to virtual function during construction [clang-analyzer-optin.cplusplus.VirtualCall,-warnings-as-errors]
      fD = SysOpen(path, O_RDONLY, 0644);
           ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:186:4: note: Taking false branch
   if (NeedsToWrite(optmode)) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:198:12: note: This constructor of an object of type 'TMemFile' has not returned when the virtual method was called
      fD = SysOpen(path, O_RDONLY, 0644);
           ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:198:12: note: Call to virtual function during construction
/home/performance-test/src_master/io/io/src/TMemFile.cxx:646:13: error: Null pointer passed as an argument to a 'nonnull' parameter [clang-analyzer-core.NonNullParamChecker,-warnings-as-errors]
            memcpy(fBlockSeek->fBuffer, buf, fBlockSeek->fSize);
            ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:672:11: note: Calling 'TMemFile::SysWriteImpl'
   return SysWriteImpl(fd,buf,len);
          ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:610:4: note: Taking false branch
   if (fExternalData) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:615:8: note: Assuming the condition is false
   if (fBlockList.fBuffer == 0) {
       ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:615:4: note: Taking false branch
   if (fBlockList.fBuffer == 0) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:620:11: note: Assuming the condition is false
      if (fBlockOffset+len <= fBlockSeek->fSize) {
          ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:620:7: note: Taking false branch
      if (fBlockOffset+len <= fBlockSeek->fSize) {
      ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:636:14: note: Assuming the condition is true
         if (!fBlockSeek->fNext) {
             ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:636:10: note: Taking true branch
         if (!fBlockSeek->fNext) {
         ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:637:13: note: Calling 'TMemBlock::CreateNext'
            fBlockSeek->CreateNext(fDefaultBlockSize);
            ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:90:16: note: Calling constructor for 'TMemBlock'
   fNext = new TMemBlock(size,this);
               ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:60:35: note: Null pointer value stored to field 'fBuffer'
   fPrevious(previous), fNext(0), fBuffer(0), fSize(0)
                                  ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:63:8: note: Assuming the condition is false
   if (size != -1) {
       ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:63:4: note: Taking false branch
   if (size != -1) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:90:16: note: Returning from constructor for 'TMemBlock'
   fNext = new TMemBlock(size,this);
               ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:637:13: note: Returning from 'TMemBlock::CreateNext'
            fBlockSeek->CreateNext(fDefaultBlockSize);
            ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:643:17: note: Assuming the condition is true
         while (len_left > fBlockSeek->fSize) {
                ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:643:10: note: Loop condition is true.  Entering loop body
         while (len_left > fBlockSeek->fSize) {
         ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:646:13: note: Null pointer passed as an argument to a 'nonnull' parameter
            memcpy(fBlockSeek->fBuffer, buf, fBlockSeek->fSize);
            ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:658:10: error: Null pointer passed as an argument to a 'nonnull' parameter [clang-analyzer-core.NonNullParamChecker,-warnings-as-errors]
         memcpy(fBlockSeek->fBuffer, buf, len_left);
         ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:672:11: note: Calling 'TMemFile::SysWriteImpl'
   return SysWriteImpl(fd,buf,len);
          ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:610:4: note: Taking false branch
   if (fExternalData) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:615:8: note: Assuming the condition is false
   if (fBlockList.fBuffer == 0) {
       ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:615:4: note: Taking false branch
   if (fBlockList.fBuffer == 0) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:620:11: note: Assuming the condition is false
      if (fBlockOffset+len <= fBlockSeek->fSize) {
          ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:620:7: note: Taking false branch
      if (fBlockOffset+len <= fBlockSeek->fSize) {
      ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:636:14: note: Assuming the condition is true
         if (!fBlockSeek->fNext) {
             ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:636:10: note: Taking true branch
         if (!fBlockSeek->fNext) {
         ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:637:13: note: Calling 'TMemBlock::CreateNext'
            fBlockSeek->CreateNext(fDefaultBlockSize);
            ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:90:16: note: Calling constructor for 'TMemBlock'
   fNext = new TMemBlock(size,this);
               ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:60:35: note: Null pointer value stored to field 'fBuffer'
   fPrevious(previous), fNext(0), fBuffer(0), fSize(0)
                                  ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:63:8: note: Assuming the condition is false
   if (size != -1) {
       ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:63:4: note: Taking false branch
   if (size != -1) {
   ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:90:16: note: Returning from constructor for 'TMemBlock'
   fNext = new TMemBlock(size,this);
               ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:637:13: note: Returning from 'TMemBlock::CreateNext'
            fBlockSeek->CreateNext(fDefaultBlockSize);
            ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:643:17: note: Assuming the condition is false
         while (len_left > fBlockSeek->fSize) {
                ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:643:10: note: Loop condition is false. Execution continues on line 657
         while (len_left > fBlockSeek->fSize) {
         ^
/home/performance-test/src_master/io/io/src/TMemFile.cxx:658:10: note: Null pointer passed as an argument to a 'nonnull' parameter
         memcpy(fBlockSeek->fBuffer, buf, len_left);
         ^
Suppressed 97 warnings (97 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
6 warnings treated as errors
```

File: io/io/inc/TMemFile.h (hunk 4 above)
Comment (vgvassilev) [reply io/io/inc/TMemFile.h]: That would mean going through several layers of indirection and relying on reflection. That seems to make the test significantly more complex to debug. I'd prefer not taking that route.

**Pull request merged**